cql3/expr/to_restriction.cc

```diff
@@ -20,15 +20,246 @@
+            [&](const unresolved_identifier& unin) -> const column_definition* { return resolve_column(unin, schema); },
+            [](const auto&) -> const column_definition* { throw std::runtime_error("Only columns can be subscripted!"); }
+        },
+        sub.val);
+
+    auto reset_processing_pk_column = defer([&ctx] () noexcept { ctx.set_processing_pk_restrictions(false); });
+    if (subscripted_col->is_partition_key()) {
+        ctx.set_processing_pk_restrictions(true);
+    }
+
+    validate_single_column_relation(subscripted_col, true, oper, rhs, schema);
+
+    const collection_type_impl* collection = dynamic_cast<const collection_type_impl*>(subscripted_col->type.get());
+    lw_shared_ptr<column_specification> key_receiver =
+        collection->make_collection_receiver(*subscripted_col->column_specification, true);
+    lw_shared_ptr<column_specification> value_receiver =
+        collection->make_collection_receiver(*subscripted_col->column_specification, true);
+
+    expression prepared_sub = prepare_expression(sub, db, schema.ks_name(), key_receiver);
+    fill_prepare_context(prepared_sub, ctx);
+
+    expression prepared_lhs = subscript{.val = column_value(subscripted_col), .sub = std::move(prepared_sub)};
+    lw_shared_ptr<column_specification> lhs_receiver = value_receiver;
+
+    lw_shared_ptr<column_specification> rhs_receiver;
+
+    if (oper == oper_t::IN) {
+        // TODO: Fix this
+        data_type lhs_type = subscripted_col->column_specification->type->underlying_type();
+        data_type rhs_receiver_type = list_type_impl::get_instance(std::move(lhs_type), false);
+        column_specification& lhs_spec = *subscripted_col->column_specification;
+        rhs_receiver = make_lw_shared<column_specification>(lhs_spec.ks_name,
+                                                            lhs_spec.cf_name,
+                                                            lhs_spec.name,
+                                                            std::move(rhs_receiver_type));
+    } else {
+        rhs_receiver = lhs_receiver;
+    }
+
+    expression prepared_rhs = prepare_expression(rhs, db, schema.ks_name(), rhs_receiver);
+    fill_prepare_context(prepared_rhs, ctx);
+
+    ::shared_ptr<restrictions::restriction> r = ::make_shared<restrictions::single_column_restriction>(*subscripted_col);
+    r->expression = binary_operator(std::move(prepared_lhs), oper, std::move(prepared_rhs), order);
+    return r;
+}
+} // anonymous namespace
+
+::shared_ptr<restrictions::restriction> to_restriction(const expression& e,
                                                        data_dictionary::database db,
                                                        schema_ptr schema,
                                                        prepare_context& ctx) {
+
+    if (is<constant>(e)) {
+        throw exceptions::invalid_request_exception("Constant value by itself is not supported as a restriction yet");
```

Comment:
I definitely want to support it, not because it is useful, but because the grammar is a lot more flexible when a subexpression can be any expression, not just one of a few supported shapes.